sw/host/opentitanlib/src/transport/hyperdebug/mod.rs

```diff
@@ -380,9 +379,24 @@ impl<T: Flavor> Transport for Hyperdebug<T> {
 
     // Crate SPI Target instance, or return one from a cache of previously created instances.
     fn spi(&self, instance: &str) -> Result<Rc<dyn Target>> {
-        let &idx = self.spi_names.get(instance).ok_or_else(|| {
-            TransportError::InvalidInstance(TransportInterfaceType::Spi, instance.to_string())
-        })?;
+        // Execute a "spiget" command to look up the numeric index corresponding to the given
+        // textual SPI instance name.
+        let mut result: Result<u8> = Err(TransportError::InvalidInstance(
+            TransportInterfaceType::Spi,
+            instance.to_string(),
+        )
+        .into());
+        let regex = Regex::new("^ +([0-9]+) ([^ ]+) ([0-9]+)")?;
+        self.inner
+            .execute_command(&format!("spiget {}", instance), |line| {
+                if let Some(captures) = regex.captures(line) {
+                    if let Ok(index) = captures.get(1).unwrap().as_str().parse() {
+                        result = Ok(index)
+                    }
+                }
+            })?;
+        let idx = result?;
```

Comment:
Based on how this works with the callback, I think execute_command should be something like:
```
fn execute_comand<F, R>(&self, cmd: &str, callback: F)  -> Result<R>
where:
    F: Fn(&str) -> Result<R>,
{
   ...
}
```

That would allow you to not have the local `mut result` and to write the execute expression as something like:

```
let idx = self.inner.execute_command(&format!(...), |line| {
    if let Some(captures) = REGEX.captures(line) {
        ...
    } else {
       Err(TransportError::<whatever>)
    }
})?;
```

Since this is more of a cleanup, its ok with me to do that in a follow-up PR.

Reply:
I know.  This execute_command() was designed to also handle commands which result in multiple lines of text output, and it invokes the given function once for each line, allowing a streaming processing (as opposed to returning a list of strings).
I am not certain this is the best design, in fact, it is becoming apparent that it is not, as it puts the burden of handling unexpected cases of a second unexpected line of output on the caller.

I think I want to make several convenience methods, one probably for executing a command that outputs nothing on success, which would return an error on any output.  Another for cases in which the caller expects exactly one line of output, which would be returned.  Possibly the latter can by further refined into expecting exactly one line of output which matches a given Regex, returning a Captures object.

I would want to look into that in a separate refactoring CL.